### ISSUE TITLE
Improve xUnit1045 to avoid firing for serializable arguments stored in `object`

### ISSUE THREAD

**Issue by JakeYallop** [xunit/xunit#2981]
### Reproduction:

```csharp
public class UnitTest1
{
    public static TheoryData<object> Data => new()
    {
        { 1 },
    };

    [Theory]
    [MemberData(nameof(Data))] // xUnit1045: The type argument object might not be serializable, which may cause Test Explorer to not enumerate individual data rows. Consider using a type that is known to be serializable.
    public void Test1(object value)
    {
        Assert.Equal(1, value);
    }
}
```

### Description
This is more of a suggested improvement rather than a bug, as the analyzer reports "may not be serializable", which is completely correct. Although my use case is specifically for primitive types, this could be extended to handle any data that implements `IXUnitSerializable`.

For more complicated data generators it may not be feasible to figure out this information.

**Comment by marcelbeeker** [comment]
I have a complex custom type witch includes other complex type. To solve this suggestion I have to make all type derived from IXunitSerializable. As a result I have to use XUnit in my API also. That's not what I want. 

For example I have this class:

public class TestDataSet
{
    public RefreshTokenDto RefreshTokenDto { get; set; }

    public RefreshTokenDtoExpectedResult ExpectedResult { get; set; }
}

To solve this I have to implement the IXunitSerializable interface for all classes which means for the TestDataSet, RefreshTokenDto  and RefreshTokenDtoExpectedResult. That's not what I want. I there a better way to solve this suggestion?

**Comment by bradwilson** [comment]
> Is there a better way to solve this suggestion?

For v3 `0.5.0-pre.27` we've added external serialization via `IXunitSerializer`: https://xunit.net/docs/getting-started/v3/custom-serialization

**Comment by ssaporito** [comment]
> > Is there a better way to solve this suggestion?
> 
> For v3 `0.5.0-pre.27` we've added external serialization via `IXunitSerializer`: https://xunit.net/docs/getting-started/v3/custom-serialization
> 
> (Note that we have not yet updated the analyzers to account for this.)

Hi @bradwilson ,

Do you know if this is still an open issue? 

The code below still triggers xUnit1045 without the IXunitSerializer, but not with it, so I think the analyzer has already been updated to account for it.  

``` C#
using Xunit.Sdk;

[assembly: RegisterXunitSerializer(typeof(CustomSerializer), typeof(UnitTest1.TestDataSet))]

public class UnitTest1
{
    public class TestDataSet
    {
        public RefreshTokenDto RefreshTokenDto { get; set; }

        public RefreshTokenDtoExpectedResult ExpectedResult { get; set; }
    }

    public class RefreshTokenDto { }

    public class RefreshTokenDtoExpectedResult { }

    public static TheoryData<TestDataSet> Data =>
    [
        new(new()),
    ];

    [Theory]
    [MemberData(nameof(Data))]
    public void Test1(TestDataSet value)
    {
        Assert.Equal(new(), value);
    }
}

public class CustomSerializer : IXunitSerializer
{
    public object Deserialize(Type type, string serializedValue) => throw new NotImplementedException();

    public bool IsSerializable(Type type, object? value, out string? failureReason)
    {
        failureReason = null;
        return true;
    }

    public string Serialize(object value) => throw new NotImplementedException();
}
```

**Comment by bradwilson** [comment]
To the best of my knowledge, the original issue (asking to improve xUnit1045 to look at the actual data, rather than the data type of the `TheoryData<>`) still exists. Your sample differs in the sense that you're using `TheoryData<TestDataSet>`, so it's evaluating the serializable-ness of `TestDataSet` vs. the original issue's `object`.

**Comment by ssaporito** [comment]
Oh, right, it only covers the issue mentioned by @marcelbeeker. I will look into how the analyzer might infer serializableness from the data if the type does not provide it, then.

**Comment by ana1250** [comment]
Hi! I’d like to work on this.

From what I understand, the current behavior flags `TheoryData<object>` because `object` is treated as `PossiblySerializable`, which leads to false positives even when the actual runtime values are serializable (e.g., primitives).

One possible approach is to ignore `System.Object` in the serializability check, similar to how certain other types are handled currently. This would reduce noise for common usage patterns, but it would also mean we no longer warn when non-serializable instances are passed via `object`.

Given that this scenario cannot be reliably analyzed at compile time (similar to arrays or other containers), would it be acceptable to favor reducing false positives by ignoring `object` entirely?

Also, I assume existing tests expecting `xUnit1045` for `object` would need to be updated accordingly please let me know if that aligns with the intended behavior.

Happy to proceed based on your guidance.

**Comment by bradwilson** [comment]
> One possible approach is to ignore `System.Object` in the serializability check, similar to how certain other types are handled currently. This would reduce noise for common usage patterns, but it would also mean we no longer warn when non-serializable instances are passed via `object`.

I don't particularly like this strategy. It doesn't matter whether using `object` for non-`object` values is common, since the entire point of this analyzer is to say "stronger typed is better". Turning it off for `object` basically turns off the primary reason it exists.

What OP is asking for is inspecting all the values to check for serializability. If this isn't really feasible than my preference would be to just close this issue.

**Comment by ana1250** [comment]
That makes sense, thanks for the clarification.

I see how allowing `object` would weaken the intent of the analyzer to encourage stronger typing. Given that, it seems the current behavior is aligned with the design goals, even if it can produce some false positives.

Happy to leave this as-is unless there’s an alternative approach you’d prefer to explore.

**Comment by JakeYallop** [comment]
My original request was to expand the behaviour - leaving as-is doesn't really improve anything. It's been a while since I've used C#, but I have written a couple of analyzers in the past. As a starting point, you could look for numeric literals, which are a `LiteralExpressionSyntax` node with a `Kind` of `NumericLiteralToken`.

As a further step, you could try to resolve constants as well:

```csharp
const int A = 1;
const int B = 2;

new TheoryData<object> { A, B }
```

I've not looked into the actual analyzer implementation myself.

**Comment by bradwilson** [comment]
> My original request was to expand the behaviour - leaving as-is doesn't really improve anything.

I agree, and I intend to leave this issue open for that reason. If someone comes up with a reasonable way of providing a better implementation when all the values can be known by inspecting the source, then I'm happy to look at that PR. (I can pretty much assure you that I won't end up doing this myself, which is mostly where you find me adding the `help wanted` tag to issues. 😁)

**Comment by bradwilson** [comment]
Available in xunit.analyzers `2.0.0-pre.48` https://xunit.net/docs/using-ci-builds
